<div dir="rtl" style="text-align:right;">

# فصل ۲: متغیرها و انواع داده در پایتون
</div>


<div dir="rtl" style="text-align:right;">

## ۲.۱ مقدمه

در فصل قبل یاد گرفتیم چطور پایتون را نصب کنیم و اولین برنامه را بنویسیم. حالا وقت آن است که با **متغیرها** و **انواع داده** آشنا شویم — این دو مفهوم، ستون فقرات هر برنامه‌ای هستند.

در این فصل یاد می‌گیریم:
- متغیر چیست و چطور تعریف می‌شود؟
- انواع داده در پایتون کدامند؟
- چطور نوع داده را بررسی و تبدیل کنیم؟
- قوانین نام‌گذاری متغیرها

---

## ۲.۲ متغیر چیست؟

یک **متغیر** (Variable) مثل یک جعبه است که می‌توانید یک مقدار را داخل آن بگذارید و بعداً از آن استفاده کنید.

در پایتون برای تعریف متغیر **نیازی به اعلان نوع** ندارید — فقط کافی است اسم بگذارید و مقدار بدهید:
</div>


<div dir="rtl" style="text-align:right;">

# ساختار کلی:
</div>


In [1]:
# variable_name = value
depth = 350
region = "Alborz"
temperature = 85.5

<div dir="rtl" style="text-align:right;">

علامت `=` در پایتون **عملگر انتساب** است، نه تساوی ریاضی. یعنی مقدار سمت راست را در متغیر سمت چپ ذخیره کن.
</div>


In [2]:
sample_name = "Granite"
sample_weight = 2.75
sample_depth = 120
is_fresh = True

print(sample_name)
print(sample_weight)
print(sample_depth)
print(is_fresh)

Granite
2.75
120
True


<div dir="rtl" style="text-align:right;">

خروجی:

Granite
2.75
120
True

---

## ۲.۳ انواع داده در پایتون

پایتون چند نوع داده پایه دارد:

| نوع داده | نام انگلیسی | مثال |
|---|---|---|
| عدد صحیح | int | 42 |
| عدد اعشاری | float | 3.14 |
| رشته متنی | str | "Basalt" |
| بولین | bool | True / False |
| خالی | NoneType | None |

---

## ۲.۴ عدد صحیح (int)

اعداد صحیح شامل اعداد مثبت، منفی و صفر هستند — **بدون اعشار**.
</div>


In [4]:
age = 25
depth = 2500
count = -10
year = 2024

print(age)
print(depth)
print(count)
print(year)

25
2500
-10
2024


<div dir="rtl" style="text-align:right;">

خروجی:

25
2500
-10

برای خوانایی بهتر اعداد بزرگ، می‌توانید از `_` به عنوان جداکننده استفاده کنید:
</div>


In [5]:
earth_mass = 5_972_000_000_000_000_000_000_000
distance = 149_597_870_700

print(earth_mass)
print(distance)

5972000000000000000000000
149597870700


<div dir="rtl" style="text-align:right;">

---

## ۲.۵ عدد اعشاری (float)

برای مقادیری که به دقت بیشتری نیاز دارند استفاده می‌شود.
</div>


In [6]:
latitude = 35.7018
longitude = 51.3890
ndvi = 0.68
pixel_size = 10.0

print(f"مختصات: {latitude} N, {longitude} E")
print(f"NDVI: {ndvi}")

مختصات: 35.7018 N, 51.389 E
NDVI: 0.68


<div dir="rtl" style="text-align:right;">

خروجی:

مختصات: 35.7018 N, 51.389 E
NDVI: 0.68

برای اعداد بسیار بزرگ یا کوچک می‌توان از نمایش علمی استفاده کرد:
</div>


In [8]:
small_value = 2.5e-6    # 2.5 * 10^-6
large_value = 1.4e9     # 1.4 * 10^9

print(small_value)
print(large_value)

2.5e-06
1400000000.0


<div dir="rtl" style="text-align:right;">

خروجی:

2.5e-06
1400000000.0

---

## ۲.۶ رشته متنی (str)

رشته‌ها برای ذخیره متن استفاده می‌شوند و باید داخل نقل‌قول قرار بگیرند.
</div>


In [9]:
rock_type = "Basalt"
formation = 'Karaj Formation'

# رشته چندخطی
description = """
سنگ‌های آتشفشانی
با سن ائوسن
"""

print(rock_type)
print(formation)

Basalt
Karaj Formation


<div dir="rtl" style="text-align:right;">

### عملیات رایج روی رشته‌ها
</div>


In [15]:
name = "Sentinel-2A"

print(len(name))                        # طول رشته
print(name.upper())                     # حروف بزرگ
print(name.lower())                     # حروف کوچک
print(name.replace("2A", "2B"))         # جایگزینی
print(name.startswith("Sentinel"))      # بررسی شروع

11
SENTINEL-2A
sentinel-2a
Sentinel-2B
True


<div dir="rtl" style="text-align:right;">

خروجی:

12
SENTINEL-2A
sentinel-2a
Sentinel-2B
True

### f-string — بهترین روش ترکیب رشته‌ها
</div>


In [19]:
satellite = "Landsat-8"
resolution = 30
date = "2024-03-15"

info = f"ماهواره: {satellite} | رزولوشن: {resolution} | تاریخ: {date}"
print(info)

ماهواره: Landsat-8 | رزولوشن: 30 | تاریخ: 2024-03-15


<div dir="rtl" style="text-align:right;">

خروجی:

ماهواره: Landsat-8 | رزولوشن: 30m | تاریخ: 2024-03-15

> **نکته:** f-string خواناترین و پیشنهادی‌ترین روش برای ترکیب متن و متغیر در پایتون است. حرف `f` را قبل از نقل‌قول بگذارید و متغیر را داخل `{}` قرار دهید.

---

## ۲.۷ بولین (bool)

فقط دو مقدار دارد: `True` یا `False`.
</div>


In [20]:
is_active = True
has_water = False
is_cloud_free = True

print(is_active)
print(has_water)
print(is_cloud_free)

True
False
True


<div dir="rtl" style="text-align:right;">

خروجی:

True
False
True

در پایتون `True` معادل `1` و `False` معادل `0` است:
</div>


In [21]:
print(True + True)    # 2
print(True + False)   # 1
print(False + False)  # 0

2
1
0


<div dir="rtl" style="text-align:right;">

---

## ۲.۸ نوع None

`None` یعنی "هیچ مقداری وجود ندارد" — برای متغیرهایی که هنوز مقدار ندارند.
</div>


In [23]:
result = None
print(result)         # None

result = 42.3
print(result)         # 42.3

None
42.3


<div dir="rtl" style="text-align:right;">

---

## ۲.۹ بررسی نوع داده با type()
</div>


In [24]:
depth = 1500
temperature = 98.6
location = "Zagros"
active = True
result = None

print(type(depth))
print(type(temperature))
print(type(location))
print(type(active))
print(type(result))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>
<class 'NoneType'>


<div dir="rtl" style="text-align:right;">

خروجی:

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>
<class 'NoneType'>

برای نمایش فقط نام نوع:
</div>


In [25]:
value = 3.14
print(type(value).__name__)    # float

float


<div dir="rtl" style="text-align:right;">

---

## ۲.۱۰ تبدیل انواع داده (Type Casting)

| تابع | عملکرد |
|---|---|
| int() | تبدیل به عدد صحیح |
| float() | تبدیل به عدد اعشاری |
| str() | تبدیل به رشته |
| bool() | تبدیل به بولین |
</div>


<div dir="rtl" style="text-align:right;">

# float به int — اعشار حذف می‌شود (گرد نمی‌شود)
</div>


In [26]:
elevation = 1847.9
print(int(elevation))       # 1847

1847


<div dir="rtl" style="text-align:right;">

# int به float
</div>


In [28]:
pixel = 128
print(float(pixel))         # 128.0

128.0


<div dir="rtl" style="text-align:right;">

# عدد به رشته
</div>


In [32]:
depth = 3500
print(str(depth) + "m")     # 3500m

3500m


<div dir="rtl" style="text-align:right;">

# رشته عددی به int
</div>


In [34]:
value = "256"
print(int(value) + 10)      # 266

266


<div dir="rtl" style="text-align:right;">

> **هشدار:** اگر رشته‌ای که تبدیل می‌کنید عدد نباشد، خطای `ValueError` می‌گیرید:
</div>


<div dir="rtl" style="text-align:right;">

# این کد خطا می‌دهد
</div>


In [35]:
x = int("Granite")
# ValueError: invalid literal for int() with base 10: 'Granite'

ValueError: invalid literal for int() with base 10: 'Granite'

<div dir="rtl" style="text-align:right;">

---

## ۲.۱۱ قوانین نام‌گذاری متغیرها

### قوانین اجباری
</div>


<div dir="rtl" style="text-align:right;">

# ✅ درست
</div>


In [36]:
rock_type = "Basalt"
depth2 = 500
_internal = True

<div dir="rtl" style="text-align:right;">

# ❌ با عدد شروع می‌شود
</div>


In [37]:
2depth = 500

SyntaxError: invalid decimal literal (3574780174.py, line 1)

<div dir="rtl" style="text-align:right;">

# ❌ فاصله دارد
</div>


In [38]:
rock type = "Basalt"

SyntaxError: invalid syntax (2249030641.py, line 1)

<div dir="rtl" style="text-align:right;">

# ❌ کاراکتر خاص
</div>


In [39]:
rock-type = "Basalt"

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (1700480680.py, line 1)

<div dir="rtl" style="text-align:right;">

### استانداردهای نام‌گذاری
</div>


<div dir="rtl" style="text-align:right;">

# snake_case — استاندارد پایتون برای متغیرها
</div>


In [40]:
rock_density = 2.7
fault_length_km = 380
image_resolution = 10

<div dir="rtl" style="text-align:right;">

# UPPER_SNAKE_CASE — برای ثابت‌ها
</div>


In [41]:
GRAVITY = 9.81
EARTH_RADIUS_KM = 6371
SPEED_OF_LIGHT = 299_792_458

<div dir="rtl" style="text-align:right;">

# نام‌های معنادار — همیشه بهتر از x و y

# ❌ بد:
</div>


In [42]:
x = 0.68
y = 0.12

<div dir="rtl" style="text-align:right;">

# ✅ خوب:
</div>


In [43]:
ndvi_value = 0.68
ndwi_value = 0.12

<div dir="rtl" style="text-align:right;">

برای دیدن کلمات رزروشده پایتون که نمی‌توانید به عنوان نام متغیر استفاده کنید:
</div>


In [44]:
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


<div dir="rtl" style="text-align:right;">

---

## ۲.۱۲ تخصیص چندگانه
</div>


<div dir="rtl" style="text-align:right;">

# یک مقدار به چند متغیر
</div>


In [46]:
band1 = band2 = band3 = 0.0
print(band1, band2, band3)

0.0 0.0 0.0


<div dir="rtl" style="text-align:right;">

# مقادیر مختلف در یک خط
</div>


In [47]:
lat, lon, elev = 35.70, 51.38, 1200
print(f"Lat: {lat}, Lon: {lon}, Elev: {elev}m")

Lat: 35.7, Lon: 51.38, Elev: 1200m


<div dir="rtl" style="text-align:right;">

# جابجایی مقادیر (Swap)
</div>


In [49]:
x = 100
y = 200
x, y = y, x
print(x, y)    # 200 100

200 100


<div dir="rtl" style="text-align:right;">

خروجی همه با هم:

0.0 0.0 0.0
Lat: 35.7, Lon: 51.38, Elev: 1200m
200 100

---

## ۲.۱۳ مثال کاربردی: پروفایل یک تصویر ماهواره‌ای
</div>


<div dir="rtl" style="text-align:right;">

# اطلاعات تصویر
</div>


In [50]:
scene_id = "LC08_L2SP_164034_20230815"
satellite = "Landsat-8"
date = "2023-08-15"
cloud_cover = 3.2
pixel_size = 30.0

<div dir="rtl" style="text-align:right;">

# مقادیر بازتابندگی
</div>


In [51]:
band_red = 0.103
band_nir = 0.312

<div dir="rtl" style="text-align:right;">

# محاسبه NDVI
</div>


In [52]:
ndvi = (band_nir - band_red) / (band_nir + band_red)

<div dir="rtl" style="text-align:right;">

# گزارش
</div>


In [53]:
print("=" * 45)
print(f"  ماهواره : {satellite}")
print(f"  تاریخ   : {date}")
print(f"  شناسه   : {scene_id}")
print(f"  پوشش ابر: {cloud_cover}%")
print(f"  رزولوشن : {pixel_size}m")
print("-" * 45)
print(f"  Red  : {band_red}")
print(f"  NIR  : {band_nir}")
print(f"  NDVI : {ndvi:.4f}")
print("=" * 45)

  ماهواره : Landsat-8
  تاریخ   : 2023-08-15
  شناسه   : LC08_L2SP_164034_20230815
  پوشش ابر: 3.2%
  رزولوشن : 30.0m
---------------------------------------------
  Red  : 0.103
  NIR  : 0.312
  NDVI : 0.5036


<div dir="rtl" style="text-align:right;">

خروجی:

=============================================
  ماهواره : Landsat-8
  تاریخ   : 2023-08-15
  شناسه   : LC08_L2SP_164034_20230815
  پوشش ابر: 3.2%
  رزولوشن : 30.0m
---------------------------------------------
  Red  : 0.103
  NIR  : 0.312
  NDVI : 0.5026
=============================================

---

## ۲.۱۴ تمرین‌های عملی

### تمرین ۱
متغیرهای زیر را تعریف کرده و نوع هر کدام را با `type()` چاپ کنید:
- نام یک کوه
- ارتفاع آن به متر
- آیا قله برفی دارد؟
- تاریخ آخرین صعود (رشته)

### تمرین ۲
دو متغیر `band_green = 0.118` و `band_nir = 0.287` تعریف کنید، شاخص NDWI را محاسبه و با f-string چاپ کنید:
</div>


<div dir="rtl" style="text-align:right;">

NDWI = (Green - NIR) / (Green + NIR)

### تمرین ۳
اطلاعات یک چاه حفاری (نام، عمق، دما، فشار، وضعیت تولید) را در متغیرهای مناسب ذخیره کرده و یک گزارش کامل با f-string چاپ کنید.

---

## ۲.۱۵ خلاصه

| مفهوم | نکته کلیدی |
|---|---|
| متغیر | فضای ذخیره‌سازی نام‌گذاری‌شده |
| int | عدد صحیح، بدون اعشار |
| float | عدد اعشاری |
| str | متن داخل نقل‌قول |
| bool | فقط True یا False |
| None | بدون مقدار |
| type() | بررسی نوع داده |
| int(), float(), str() | تبدیل نوع داده |
| f-string | بهترین روش ترکیب متن و متغیر |
| snake_case | استاندارد نام‌گذاری پایتون |

---

*فصل بعدی: عملگرها و عبارات در پایتون*
</div>
